### ISSUE TITLE
CMake options without prefix 

### ISSUE THREAD

**Issue by dutkalex** [cburstedde/p4est#302]
**Description**
The ``mpi``, ``openmp``, ``json``, ``enable-file-deprecated``, ``enable_p6est``, ``enable_p8est``, ``vtk_binary`` and ``zlib`` CMake options can easily collide with other projects. I think that it is a significant problem especially for the ``mpi``, ``openmp``, ``json``and ``zlib``, which should be renamed to something like ``P4EST_ENABLE_MPI``, ``P4EST_ENABLE_OPENMP``, ``P4EST_ENABLE_JSON`` and ``P4EST_ENABLE_ZLIB`` respectively. It would also disambiguate the exact behavior associated to each option. This is a very low-hanging fruit in terms of hardening of the CMake build system, and improving the interoperability with external projects. I can easily propose a small PR to fix this.

**To Reproduce**
``` 
mkdir build
cd build
ccmake ..
```

**Comment by cburstedde** [comment]
> **Description**
> The ``mpi``, ``openmp``, ``json``, ``enable-file-deprecated``, ``enable_p6est``, ``enable_p8est``, ``vtk_binary`` and ``zlib`` CMake options can easily collide with other projects. I think that it is a significant problem especially for the ``mpi``, ``openmp``, ``json``and ``zlib``, which should be renamed to something like ``P4EST_ENABLE_MPI``, ``P4EST_ENABLE_OPENMP``, ``P4EST_ENABLE_JSON`` and ``P4EST_ENABLE_ZLIB`` respectively. It would also disambiguate the exact behavior associated to each option. 
> 
> **To Reproduce**
> ``` 
>  mkdir build; cd build
> ccmake ..
> ```
> 
> This is a very low-hanging fruit in terms of hardening of the CMake build system. I can easily propose a small PR to fix this.

Thanks; we've recently stumbled upon this phenomenon of having two variables
for the same setting.  These should be unified to the variable names used in
autoconf and as #define, that is, go for the P4EST_ uppercase versions.

While at it, we may symmetrize cmake to add P4EST_ENABLE_BUILD_2D on top of
P4EST_ENABLE_BUILD_3D and P4EST_ENABLE_P6EST.  Please go ahead if you like.

**Comment by dutkalex** [comment]
> Please go ahead if you like.

Should I branch for the master or the develop?

**Comment by dutkalex** [comment]
Some problems originate from the `sc` library, so fixing this will require a PR on the `sc` repo too.

**Comment by cburstedde** [comment]
> Some problems originate from the `sc` library, so fixing this will require a PR on the `sc` repo too.

Yes, both libraries would need similar treatment.  If it would be possible
to reuse some CMake code between them, that would be great.  It would be
good if the Presets structure remains functional, since third parties rely
on them for their builds, especially about setting debug and mpi modes.

Please base of develop, and thanks!  It may be that we address sc first and
then I'd update the submodule in p4est, before the PR can be posted there --
but it may also be that the two PRs interdepend, please comment on this.

**Comment by dutkalex** [comment]
Yes I agree. I will start investigating the question. It is not obvious to me at the time whether the two can be decoupled or not...

**Comment by dutkalex** [comment]
The PR for `libsc` is ready. Waiting on your feedback for integration and see what needs to be adapted on the p4est side now...

**Comment by cburstedde** [comment]
> The PR for `libsc` is ready. Waiting on your feedback for integration. Once this is merged in the `develop`, I'll be able to see what needs to be adapted on the p4est side...

This sounds good, thanks for keeping at it!

**Comment by cburstedde** [comment]
Once we get the libsc CMake/CI to go through, I'll update the sc submodule in p4est. Then we may proceed along the same lines. We can use the CMake variables `P4EST_ENABLE_*` throughout and leave the one -Dmpi shortcut as with libsc.

**Comment by cburstedde** [comment]
The sc updates are merged. Would you see some room for yourself to align the p4est behavior to sc (assuming the sc behavior is to your liking/consistent, please let us know)?

**Comment by CsatiZoltan** [comment]
Any update on this issue? I use p4est as a CMake subproject and I would like to know the future of CMake in p4est/sc. Thank you.

**Comment by cburstedde** [comment]
On the current feature-distro branches, we are removing these settings from pkgconfig, reducing it to the standard minimum. Would the problem go away then?

**Comment by CsatiZoltan** [comment]
I do not use pkgconfig, but the CMake build system of p4est. There is no problem per se on my side, I am just interested in the upcoming CMake changes to p4est. I build p4est as part of my library with the options `-Denable_p6est=ON -Denable_p8est=ON`. Then I stumbled upon this GitHub issue that you plan to rename some CMake flags. If these flags are renamed, my build instructions need to be updated. That's why I asked how p4est CMake will evolve.

**Comment by cburstedde** [comment]
> I do not use pkgconfig, but the CMake build system of p4est. There is no problem per se on my side, I am just interested in the upcoming CMake changes to p4est. I build p4est as part of my library with the options `-Denable_p6est=ON -Denable_p8est=ON`. Then I stumbled upon this GitHub issue that you plan to rename some CMake flags. If these flags are renamed, my build instructions need to be updated. That's why I asked how p4est CMake will evolve.

Thanks for the context!  We're aiming for long-term stable configuration.
The uppercase prefixed versions may be better.  We're currently a bit
distracted but I'd be looking into this again by the end of the month.

**Comment by CsatiZoltan** [comment]
Great, thank you. If you introduce changes in the CMake variables, please indicate it in this issue so that I get a notification.

**Comment by cburstedde** [comment]
Thanks for the note in general! We should do the following far CMake:
 - Eliminate BUILD_TESTING since SC_BUILD_TESTING and P4EST_BUILD_TESTING exist.
 - Prefix the GIT_VERSION_GEN variable.
 - Rename VALGRIND to P4EST_TEST_WITH_VALGRIND (this is ok for libsc)
 - Rename enable-file-deprecated to P4EST_ENABLE_FILE_DEPRECATED
 - Rewrite enable_p6est and enable_p8est to use P4EST_ENABLE_2D, P4EST_ENABLE_3D and P4EST_ENABLE_P6EST (all with default ON), with the same semantics as automake
 - Rename vtk_binary to P4EST_ENABLE_VTK_BINARY
 - Prefix jansson_DIR with library and make it all uppercase.

**Comment by cburstedde** [comment]
Add another point:
 - Currently, CMake puts .tarball_version into the source directory. Build directory would be preferred.

We won't fix all these for the upcoming release but will keep them on the list for the one after that.

**Comment by cburstedde** [comment]
Ah, and this one:
 - The CMake default prefix is /usr/local. Change to build_directory/local to agree with the autotools.
